client/labml/internal/visualizations/density.py

```diff
@@ -0,0 +1,282 @@
+from typing import List, overload, Optional, Union
+import altair as alt
+import numpy as np
+import torch
+
+BASIS_POINTS = [
+    0,
+    6.68,
+    15.87,
+    30.85,
+    50.00,
+    69.15,
+    84.13,
+    93.32,
+    100.00
+]
+
+
+def _remove_names_prefix(names: List[Union[str, List[str]]]) -> List[str]:
+    if len(names) == 0:
+        return []
+
+    if isinstance(names[0], list):
+        common = names[0]
+    else:
+        common = None
+
+    for n in names:
+        if common is None:
+            break
+        if not isinstance(n, list):
+            common = None
+        merge = []
+        for x, y in zip(common, n):
+            if x != y:
+                merge.append(None)
+            else:
+                merge.append(x)
+        common = merge
+
+    res = []
+    for n in names:
+        if isinstance(n, list):
+            if common is not None:
+                n = [p for i, p in enumerate(n) if i > len(common) or p != common[i]]
+            n = '-'.join(n)
+
+        res.append(n)
+
+    return res
+
+
+def _data_to_table(series, names, step, levels=5):
+    table = []
+
+    for s in range(len(series)):
+        data = series[s]
+        name = names[s]
+
+        try:
+            import torch
+        except ImportError:
+            torch = None
+
+        if torch is not None and isinstance(data, torch.Tensor):
+            data = data.detach().cpu().numpy()
+
+        for i in range(data.shape[0]):
+            if len(data.shape) == 2:
+                if step is not None:
+                    row = {'series': name, 'step': step[i]}
+                else:
+                    row = {'series': name, 'step': i}
+
```

Comment:
Doing percentile calculations here. I think keeping here is confusing better to get it out.

Why i did it
- until now series can be a tensor or nd array conversion is done here.
- I guess to manipulate it have to move the conversion up the hierarchy